server: embed the pprof ui

tbg opened the pull request:
Now that pprof has a web ui, we can just embed it into the running
server directly.

It turns out that for the most part, pprof has the right abstractions
built in, and so the code for this came out fairly clean. I needed a
few upstream modifications which are currently pending; it's looking
good that they'll be merged:

https://github.com/google/pprof/pulls/tschottdorf

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24145)
<!-- Reviewable:end -->

bdarnell commented:
Sweet! It's about time pprof got a built-in UI. I don't think there's any issue with exposing this (but we should ensure that the new handler has the same access controls we use for `/debug/pprof/profile`)

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24145#-:-L8DpE4bTBTeNPdVO-OF:b-zcimba)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I’m a huge fan of this! Bravo!

tschottdorf commented:
Big update! Not 100% ready to merge yet, but 100% ready to use.

nvanbenschoten commented:
This is incredible, I feel like I'm in the future!

A few quick things I noticed while playing around are:
- there's no link to the ui on the `/debug` page. That's a problem because this should be front and center!
- related to [this TODO](https://github.com/cockroachdb/cockroach/pull/24145/files#diff-596c9e506e7d72b52df632243b764f44R30), the `Search regexp` functionality doesn't seem to work at the moment - it redirects to `/?f=<query>`. It must be in need of some HTML rewriting love.

vilterp commented:
Visual nit from your demo: would be nice to put "raw" and "UI" links next to each other on the debug index page, to avoid taking up more vertical space (it's pretty long). Can provide some JSX for how to do this.

couchand commented:
Really neat stuff.  Getting rid of the link rewriting would be nice (I made a suggestion for a lighter-weight option if the upstream change doesn't move quickly), and I have a question about the nested profile server.

---

Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[Gopkg.lock, line 575 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8gvPEX3UwKnYKD2F7b:-L8gvPEX3UwKnYKD2F7c:b-flcoeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/Gopkg.lock#L575)):*
> ```
>     "internal/symbolz",
>     "profile",
>     "third_party/d3",
> ```

Too bad that this needs to pull in its own copy of d3.  The binary will be a bit bigger just because it would be nontrivial to share our copy with pprof, which is annoying.

---

*[pkg/server/debug/pprofui/script_injector.go, line 28 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8gyx6a-zYO_x3bZDUI:-L8gyx6a-zYO_x3bZDUJ:b-rcvhny) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/script_injector.go#L28)):*
> ```Go
> // emptied via WriteReplaced. It's used to massage the links in the
> // pprof web ui which unfortunately assume that the website is served
> // up at "/".
> ```

You could also just inject a `<base>` tag into the page's head and the browser will rewrite all URLs for you.

---

*[pkg/server/debug/pprofui/server.go, line 67 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h58TvE0mWQ6TIeYnM:-L8h58TvE0mWQ6TIeYnN:bcj20va) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L67)):*
> ```Go
> 
> 	// Register endpoints for heap, block, threadcreate, etc.
> 	for _, p := range runtimepprof.Profiles() {
> ```

This seems like a bit of a roundabout way to setup this server.  Can we chat about this in meatspace?

---

*[pkg/server/debug/pprofui/server.go, line 82 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h0J0h4vp_GmUnSzuV:-L8h0J0h4vp_GmUnSzuW:b-h94wcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L82)):*
> ```Go
> 	fetchHandler, pat := profileMux.Handler(r)
> 	if pat == "" {
> 		w.Write([]byte("unknown profile type for " + r.URL.String()))
> ```

should this perhaps set the status code to 404?

---

*[pkg/server/debug/pprofui/server.go, line 127 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h8J5Q1jPsyUU2AOSF:-L8h8J5Q1jPsyUU2AOSG:b-8eja1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L127)):*
> ```Go
> 	f := func(src string, duration, timeout time.Duration) (*profile.Profile, string, error) {
> 		rec := httptest.NewRecorder()
> 		req, err := http.NewRequest(
> ```

does this even need to go through `http.Request`/`http.Server`?  couldn't we just directly map the `fetch` params to the `profile.Profile`?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L8h8hvDDDPC0fZBUiXT:bu7q6ie)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok, I'm officially taking this out of "wip" status. I'm waiting for one upstream PR to merge and then we're good to go. The code unfortunately has changed a lot so I'll need a fresh review.

---

Review status: 2 of 11 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Gopkg.lock, line 575 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8gvPEX3UwKnYKD2F7b:-L8nwDq3En0lq5v9CBag:bpwy6jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/Gopkg.lock#L575)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Too bad that this needs to pull in its own copy of d3.  The binary will be a bit bigger just because it would be nontrivial to share our copy with pprof, which is annoying.
</blockquote></details>

Yeah, not sure what to do about this.

---

*[pkg/server/debug/pprofui/server.go, line 67 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h58TvE0mWQ6TIeYnM:-L8nwG3jEzN6d9hKVidn:br24pkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L67)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This seems like a bit of a roundabout way to setup this server.  Can we chat about this in meatspace?
</blockquote></details>

Take a look at the rewrite, I think you'll like it a whole lot more.

---

*[pkg/server/debug/pprofui/server.go, line 82 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h0J0h4vp_GmUnSzuV:-L8nwIyVEDmAP8t3cBVr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L82)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

should this perhaps set the status code to 404?
</blockquote></details>

Done.

---

*[pkg/server/debug/pprofui/server.go, line 127 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h8J5Q1jPsyUU2AOSF:-L8nwKAR5s7jGNb2UP7-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L127)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

does this even need to go through `http.Request`/`http.Server`?  couldn't we just directly map the `fetch` params to the `profile.Profile`?
</blockquote></details>

Done.

---

*[pkg/server/debug/pprofui/script_injector.go, line 28 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8gyx6a-zYO_x3bZDUI:-L8nwL2f1AFavMPhWCU6:by4ybnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/script_injector.go#L28)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

You could also just inject a `<base>` tag into the page's head and the browser will rewrite all URLs for you.
</blockquote></details>

Fixed upstream (I hope they merge it).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L8nwO7d4CV3wB3dsgEY:b-7f9px9)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Does this need graphviz installed or is it a new d3-based implementation of the layout? If it needs graphviz we should probably add it to the docker image. 

---

Reviewed 2 of 9 files at r2, 9 of 10 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L8p7HFGAYI1nFfP4evQ:b6ezkqr)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Ah, as long as the flamegraph is available dot probably isn't necessary. Let's leave `dot` out then (I'd probably even default to the flamegraph version when dot is available just for consistency, although I say this without having seriously used the flamegraph mode yet)

nvanbenschoten commented:
Reviewed 2 of 9 files at r2, 7 of 10 files at r3, 2 of 3 files at r4.
Review status: 11 of 12 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/server/debug/server.go, line 131 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgNH60L2C90UBflI6:-L8rgNH60L2C90UBflI7:b-wkabxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/server.go#L131)):*
> ```Go
> 	}
> 
> 	ps := pprofui.NewServer(pprofui.NewMemStorage(1, 0))
> ```

Why put this after the Server creation instead of continuing to `return &Server`?

---

*[pkg/server/debug/pprofui/fakeflags.go, line 19 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgaui4RXaBv7lCQQI:-L8rgaui4RXaBv7lCQQJ:b-5p62y0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/fakeflags.go#L19)):*
> ```Go
> import "github.com/spf13/pflag"
> 
> type pprofFlags struct {
> ```

Please add a comment explaining what this is used for and why it's necessary.

---

*[pkg/server/debug/pprofui/response_writer.go, line 22 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgj1OEknOMp2EVcHB:-L8rgj1OEknOMp2EVcHC:bc3hwxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/response_writer.go#L22)):*
> ```Go
> )
> 
> type responseBridge struct {
> ```

Same thing here - a comment would be nice.

---

*[pkg/server/debug/pprofui/server.go, line 42 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rh4KQBSNFv9WlBWJJ:-L8rh4KR76GAB7-W1PDk:b-zi5yke) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L42)):*
> ```Go
> // A Server serves up the pprof web ui. A request to /<profiletype>
> // generates a profile of the desired type and redirects to the UI for
> // it at /<profiletype>/<id>. Valid profile types at the time of
> ```

Am I reading this correctly that `/<profiletype>` now redirects instead of returning the profile? Can we still download profiles?

---

*[pkg/server/debug/pprofui/server.go, line 67 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rhVLX0ry4dQUPTPA6:-L8rhVLX0ry4dQUPTPA7:bd96sey) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L67)):*
> ```Go
> 
> 	s.profileTypes = map[string]http.HandlerFunc{
> 		"profile": func(w http.ResponseWriter, r *http.Request) {
> ```

Make a note that this is a semi-custom profile type. Or is this not part of `pprof.Profiles()`?

---

*[pkg/server/debug/pprofui/server.go, line 94 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8ribrOCOVT6GCdoiC6:-L8ribrOCOVT6GCdoiC7:b-9y01gi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L94)):*
> ```Go
> func (s *Server) parsePath(reqPath string) (profType string, id string, remainingPath string) {
> 	parts := strings.Split(path.Clean(reqPath), "/")
> 	if path.IsAbs(reqPath) {
> ```

Explain what this is about. Is it throwing away an empty beginning?

---

*[pkg/server/debug/pprofui/server.go, line 152 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rnvBb4l5ENGlqXFNW:-L8rnvBb4l5ENGlqXFNX:b-dmpibm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L152)):*
> ```Go
> 		}
> 
> 		if err := driver.PProf(&driver.Options{
> ```

This is kind of crazy, so add a comment here explaining what's going on and why we're dealing with flags.

---

*[pkg/server/debug/pprofui/ui.go, line 29 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rpeLXBcykvDQKHZo6:-L8rpeLXBcykvDQKHZo7:b-e9h7mq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/ui.go#L29)):*
> ```Go
> }
> 
> type fakeUI struct{}
> ```

Make a note that this is implementing `driver.UI`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 11 of 12 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/server/debug/pprofui/server.go, line 67 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rhVLX0ry4dQUPTPA6:-L8rwjmb88YD3cjiRpfQ:b-cqgskc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L67)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Make a note that this is a semi-custom profile type. Or is this not part of `pprof.Profiles()`?
</blockquote></details>

The latter. CPU profiles are a special snowflake.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell done.

Added a (very basic) test. I think we're mostly waiting for upstream from this point on.

---

Review status: 8 of 13 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/server/debug/server.go, line 131 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgNH60L2C90UBflI6:-L8rwRhPAoAoxdTKhKeR:bt4mzr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/server.go#L131)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why put this after the Server creation instead of continuing to `return &Server`?
</blockquote></details>

I find this more legible. The line got pretty long and was awkward to split.

---

*[pkg/server/debug/pprofui/fakeflags.go, line 19 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgaui4RXaBv7lCQQI:-L8rxEEB3Rcva5EDbOY6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/fakeflags.go#L19)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Please add a comment explaining what this is used for and why it's necessary.
</blockquote></details>

Done.

---

*[pkg/server/debug/pprofui/response_writer.go, line 22 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgj1OEknOMp2EVcHB:-L8rx_cR4vXvCyoUochu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/response_writer.go#L22)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same thing here - a comment would be nice.
</blockquote></details>

Done.

---

*[pkg/server/debug/pprofui/server.go, line 42 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rh4KQBSNFv9WlBWJJ:-L8rxajuEu2a2Mbuiarj:bktjnwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L42)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Am I reading this correctly that `/<profiletype>` now redirects instead of returning the profile? Can we still download profiles?
</blockquote></details>

The server is anchored at `/debug/pprof/ui/`, so downloading the profiles is still just a matter of hitting `/debug/pprof/<profileName>`.

---

*[pkg/server/debug/pprofui/server.go, line 67 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rhVLX0ry4dQUPTPA6:-L8rxioJCBn3SX_d11W0:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L67)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

The latter. CPU profiles are a special snowflake.
</blockquote></details>

Added a comment.

---

*[pkg/server/debug/pprofui/server.go, line 94 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8ribrOCOVT6GCdoiC6:-L8rz5ie8qF64L6ttar-:b1pwz70) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L94)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Explain what this is about. Is it throwing away an empty beginning?
</blockquote></details>

`reqPath` can be relative in theory (would be surprised to see it in practice). We just want to unify the two. I rewrote this, it was a bit clunky.

---

*[pkg/server/debug/pprofui/server.go, line 152 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rnvBb4l5ENGlqXFNW:-L8s-Fqm1anRCghLiFKF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L152)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is kind of crazy, so add a comment here explaining what's going on and why we're dealing with flags.
</blockquote></details>

Done.

---

*[pkg/server/debug/pprofui/ui.go, line 29 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rpeLXBcykvDQKHZo6:-L8s-ODT3RoHIwxBOmZB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/ui.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Make a note that this is implementing `driver.UI`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L8s3z0R-hE7KrRmpGiS:bq009h4)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> thanks for adding a bit more commentary. I'm still undecided on whether this is an example of masterful craftsmanship or downright abuse of Go. Either way, profiling just got a lot easier in Cockroach!

---

Reviewed 1 of 5 files at r5.
Review status: 9 of 13 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/server/debug/server.go, line 131 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgNH60L2C90UBflI6:-L8sPyUR1RhyqBD5bGUI:bwzyf61) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/server.go#L131)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I find this more legible. The line got pretty long and was awkward to split.
</blockquote></details>

What line got long? You're not using `s` in either of these, just `mux`.

---

*[pkg/server/debug/pprofui/server.go, line 42 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rh4KQBSNFv9WlBWJJ:-L8sQEhR0wY6jr8m6626:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L42)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The server is anchored at `/debug/pprof/ui/`, so downloading the profiles is still just a matter of hitting `/debug/pprof/<profileName>`.
</blockquote></details>

:+1: 

---

*[pkg/server/debug/pprofui/server.go, line 67 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rhVLX0ry4dQUPTPA6:-L8sQMeg2zDHvbm1obh6:b2nhoia) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/pprofui/server.go#L67)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Added a comment.
</blockquote></details>

TIL

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L8sQdkZ8GkbXHuBWCjA:b762z9k)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 8 of 13 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/server/debug/server.go, line 131 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8rgNH60L2C90UBflI6:-L8sfi2l8PH0L2-d9Y2i:b-uxsrud) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3f6cadd1aaae6352813ed65e81e3e44f33987f/pkg/server/debug/server.go#L131)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What line got long? You're not using `s` in either of these, just `mux`.
</blockquote></details>

Oh sorry, totally didn't get your comment. Done!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145)*
<!-- Sent from Reviewable.io -->

couchand commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> good stuff!

---

Reviewed 5 of 10 files at r3, 2 of 3 files at r4, 5 of 5 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[Gopkg.lock, line 575 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8gvPEX3UwKnYKD2F7b:-L95wHtOAbS0zN8Dzfvq:bbeaa5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/Gopkg.lock#L575)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, not sure what to do about this.
</blockquote></details>

Don't worry about it, it's not really a big deal!

---

*[pkg/server/debug/pprofui/response_writer.go, line 30 at r6](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L95x0gjDiCJPxhPMq3x:-L95x0gjDiCJPxhPMq3y:b-7je4b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c8c8286da0123b62ee9ad81d8110678ce8df609/pkg/server/debug/pprofui/response_writer.go#L30)):*
> ```Go
> }
> 
> var _ http.ResponseWriter = &responseBridge{}
> ```

slick

---

*[pkg/server/debug/pprofui/server.go, line 67 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L8h58TvE0mWQ6TIeYnM:-L95xUHvBDUQpJE2CS77:b-87u3pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f085b30df5621eff3f13bf670e1fb10ab1519024/pkg/server/debug/pprofui/server.go#L67)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Take a look at the rewrite, I think you'll like it a whole lot more.
</blockquote></details>

I do!

---

*[pkg/server/debug/pprofui/server.go, line 94 at r6](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L95xrO456i-k6eRQ0VK:-L95xrO456i-k6eRQ0VL:b-3mzdlp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c8c8286da0123b62ee9ad81d8110678ce8df609/pkg/server/debug/pprofui/server.go#L94)):*
> ```Go
> 	switch len(parts) {
> 	case 0:
> 		return "", "", ""
> ```

I can't imagine that it actually matters, but for consistency should the third return value be `"/"`?

---

*[pkg/server/debug/pprofui/server.go, line 112 at r6](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L95yFnEA3NB9XbqG879:-L95yFnEA3NB9XbqG87A:b-k3fh0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c8c8286da0123b62ee9ad81d8110678ce8df609/pkg/server/debug/pprofui/server.go#L112)):*
> ```Go
> 		}
> 		sort.Strings(names)
> 		msg := fmt.Sprintf("Try %s for either of %s", path.Join(r.RequestURI, "<profileName>"), strings.Join(names, ", "))
> ```

nit: "...for one of..."

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L96-IA05nyWqhkHeDdS:bl96izl)*
<!-- Sent from Reviewable.io -->

tbg committed:
server: embed the pprof ui

Now that pprof has a web ui, we can just embed it into the running
server directly.

It turns out that for the most part, pprof has the right abstractions
built in, and so the code for this came out fairly clean. I needed a
few upstream modifications which are currently pending; it's looking
good that they'll be merged:

https://github.com/google/pprof/pulls/tschottdorf

Release note: None

tschottdorf commented:
TFTRs! Upstream merged my PR, so the merge train is coming :train: 

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/server/debug/pprofui/server.go, line 94 at r6](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L95xrO456i-k6eRQ0VK:-L96LQbWAmuREWAhDwMg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c8c8286da0123b62ee9ad81d8110678ce8df609/pkg/server/debug/pprofui/server.go#L94)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I can't imagine that it actually matters, but for consistency should the third return value be `"/"`?
</blockquote></details>

Done.

---

*[pkg/server/debug/pprofui/server.go, line 112 at r6](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-L95yFnEA3NB9XbqG879:-L96LR0IE8jBM7sHyyQV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c8c8286da0123b62ee9ad81d8110678ce8df609/pkg/server/debug/pprofui/server.go#L112)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

nit: "...for one of..."
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24145#-:-L96L75S8Uco-vgDifdu:b-4u887i)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
bors r=nvanbenschoten

couchand commented:
Bors got interrupted by a non-bors merge.

bors r+

couchand commented:
bors r+

tschottdorf commented:
Thanks @couchand!
